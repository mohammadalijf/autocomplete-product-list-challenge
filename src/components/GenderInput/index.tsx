import { FormControl, InputLabel, Select } from "@material-ui/core";
import { SelectInputProps } from "@material-ui/core/Select/SelectInput";
import { FunctionComponent } from "react";
import {
  GenderInputOption,
  GenderInputProps,
  GenderInputChangeEvent,
} from "./props";

/**
 * Component for Choosing gender using MUI outlined native Select
 */
const GenderInput: FunctionComponent<GenderInputProps> = ({
  onChange,
  value,
  ...props
}) => {
  //#region

  const changeHandler: SelectInputProps["onChange"] = (event) => {
    onChange?.(event.target.value as GenderInputOption);
  };

  //#endregion
  return (
    <FormControl variant="outlined" fullWidth={props.fullWidth}>
      <InputLabel id="gender-input-label">Gender</InputLabel>
      <Select
        {...props}
        native
        labelId="gender-input-label"
        id="gender-input"
        label="Gender"
        value={value}
        inputProps={{ ...props.inputProps, "data-testid": "genderInput" }}
        onChange={changeHandler}
      >
        <option aria-label="None" />
        <option value={GenderInputOption.Unisex}>Unisex</option>
        <option value={GenderInputOption.Female}>Female</option>
        <option value={GenderInputOption.Male}>Male</option>
      </Select>
    </FormControl>
  );
};

export default GenderInput;
export type { GenderInputProps, GenderInputChangeEvent };
export { GenderInputOption };
