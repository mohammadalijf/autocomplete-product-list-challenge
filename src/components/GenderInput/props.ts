import { SelectProps } from "@material-ui/core";
import { CSVProductGender } from "../../services/productsAPI";

/**
 * GenderInput available options
 */
export { CSVProductGender as GenderInputOption };

export interface IGenderInputProps {
  /**
   * change handler for option value
   */
  onChange?: GenderInputChangeEvent;
  /**
   * selected value for option
   */
  value?: CSVProductGender;
}

/**
 * Event handler for GenderInput when option value changes
 * @param option: selected GenderInputOption
 */
export type GenderInputChangeEvent = (option: CSVProductGender) => void;

/**
 * GenderInputs' props mixed with SelectProps
 */
export type GenderInputProps = IGenderInputProps &
  Omit<SelectProps, "onChange" | "value">;
