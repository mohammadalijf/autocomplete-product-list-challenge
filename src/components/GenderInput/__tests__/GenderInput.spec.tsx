import { cleanup, render, RenderResult } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import GenderInput, { GenderInputOption } from "..";

describe("<GenderInput />", () => {
  afterEach(cleanup);

  describe("when displaying input", () => {
    let genderInput: RenderResult;
    const value = GenderInputOption.Female;
    beforeEach(async () => {
      genderInput = render(<GenderInput value={value} />);
    });

    it("renders correctly", async () => {
      expect(genderInput.asFragment()).toMatchSnapshot();
    });

    it("have all the GenderInputOptions", async () => {
      Object.keys(GenderInputOption).forEach((option) => {
        expect(genderInput.getByTestId("genderInput")).toHaveTextContent(
          option
        );
      });
    });

    it("value is correct", () => {
      expect(genderInput.getByTestId("genderInput")).toHaveValue(value);
    });
  });

  describe("when changing value", () => {
    let genderInput: RenderResult;
    let onChange: jest.Mock;
    beforeEach(async () => {
      onChange = jest.fn().mockName("onChange");
      genderInput = render(<GenderInput onChange={onChange} />);
    });

    it("calls the onChange handler with proper GenderInputOption", async () => {
      userEvent.selectOptions(
        genderInput.getByTestId("genderInput"),
        GenderInputOption.Female
      );
      expect(onChange).toHaveBeenCalledWith(GenderInputOption.Female);
    });
  });
});
