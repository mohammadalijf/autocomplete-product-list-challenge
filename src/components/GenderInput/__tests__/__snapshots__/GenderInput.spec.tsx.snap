// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GenderInput /> when displaying input renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
      data-shrink="true"
      id="gender-input-label"
    >
      Gender
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
    >
      <select
        aria-invalid="false"
        class="MuiSelect-root MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
        data-testid="genderInput"
        id="gender-input"
      >
        <option
          aria-label="None"
        />
        <option
          value="unisex"
        >
          Unisex
        </option>
        <option
          value="female"
        >
          Female
        </option>
        <option
          value="male"
        >
          Male
        </option>
      </select>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-3 PrivateNotchedOutline-legendNotched-4"
        >
          <span>
            Gender
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
